File: ios/Sources/Protocol.swift
import Foundation

/// The wire protocol, mirroring src/bridge/protocol.ts.
///
/// This file and that one have to agree exactly. If you change one, change the
/// other in the same commit — the seam between them is the whole reason the
/// backend never had to learn what a pair of glasses is.
enum BinaryTag: UInt8 {
    /// device -> server: 16 kHz mono s16le microphone audio
    case micPcm = 0x01
    /// device -> server: JPEG still, answering a captureFrame request
    case frameJpeg = 0x02
    /// server -> device: audio to play
    case ttsAudio = 0x10
}

/// Audio format for the whole pipeline. Fixed on both sides; the client
/// resamples to meet it rather than negotiating.
enum AudioFormat {
    static let sampleRate: Double = 16_000
    static let channels: UInt32 = 1
}

// MARK: - device -> server

/// Hand-rolled rather than Codable because every message has a different shape
/// and a `t`-tagged enum encoder is more ceremony than the six cases justify.
enum ClientMessage {
    case hello(label: String, frames: Bool, battery: Double?)
    case wake(at: Int64)
    case endOfSpeech(at: Int64)
    case bargeIn(at: Int64)
    case transcript(text: String, final: Bool, at: Int64)
    case frameError(reason: String)
    case battery(level: Double)
    case playbackDone(interrupted: Bool)
    case pong(at: Int64)

    var json: [String: Any] {
        switch self {
        case let .hello(label, frames, battery):
            return [
                "t": "hello",
                "label": label,
                "capabilities": [
                    "audioIn": true,
                    "audioOut": true,
                    "frames": frames,
                    "display": true,
                ],
                "battery": battery as Any,
                // No wake word yet: the talk button is the trigger. Porcupine
                // needs an access key and a binary dependency, and the first
                // build wants neither.
                "wake": "push-to-talk",
            ]
        case let .wake(at):
            return ["t": "wake", "at": at, "source": "push-to-talk"]
        case let .endOfSpeech(at):
            return ["t": "endOfSpeech", "at": at]
        case let .bargeIn(at):
            return ["t": "bargeIn", "at": at]
        case let .transcript(text, final, at):
            return ["t": "transcript", "text": text, "final": final, "at": at]
        case let .frameError(reason):
            return ["t": "frameError", "reason": reason]
        case let .battery(level):
            return ["t": "battery", "level": level]
        case let .playbackDone(interrupted):
            return ["t": "playbackDone", "interrupted": interrupted]
        case let .pong(at):
            return ["t": "pong", "at": at]
        }
    }
}

// MARK: - server -> device

enum ServerMessage {
    case ready(personaName: String, greeting: String)
    case captureFrame(reason: String)
    case speakBegin(mime: String, utteranceId: String)
    case speakEnd(utteranceId: String)
    case speakCancel(utteranceId: String)
    case state(phase: String, detail: String?)
    case heard(text: String, final: Bool)
    case said(text: String)
    case notice(level: String, text: String)
    case ping(at: Int64)
    /// Anything this build does not know about. Ignored rather than fatal, so an
    /// older app keeps working against a newer server.
    case unknown(String)

    static func parse(_ data: Data) -> ServerMessage? {
        guard
            let object = try? JSONSerialization.jsonObject(with: data),
            let dict = object as? [String: Any],
            let t = dict["t"] as? String
        else { return nil }

        switch t {
        case "ready":
            let persona = dict["persona"] as? [String: Any] ?? [:]
            return .ready(
                personaName: persona["name"] as? String ?? "Jarvis",
                greeting: persona["greeting"] as? String ?? ""
            )
        case "captureFrame":
            return .captureFrame(reason: dict["reason"] as? String ?? "")
        case "speakBegin":
            return .speakBegin(
                mime: dict["mime"] as? String ?? "audio/mpeg",
                utteranceId: dict["utteranceId"] as? String ?? ""
            )
        case "speakEnd":
            return .speakEnd(utteranceId: dict["utteranceId"] as? String ?? "")
        case "speakCancel":
            return .speakCancel(utteranceId: dict["utteranceId"] as? String ?? "")
        case "state":
            return .state(
                phase: dict["phase"] as? String ?? "idle",
                detail: dict["detail"] as? String
            )
        case "heard":
            return .heard(
                text: dict["text"] as? String ?? "",
                final: dict["final"] as? Bool ?? false
            )
        case "said":
            return .said(text: dict["text"] as? String ?? "")
        case "notice":
            return .notice(
                level: dict["level"] as? String ?? "info",
                text: dict["text"] as? String ?? ""
            )
        case "ping":
            return .ping(at: (dict["at"] as? NSNumber)?.int64Value ?? 0)
        default:
            return .unknown(t)
        }
    }
}

func nowMillis() -> Int64 { Int64(Date().timeIntervalSince1970 * 1000) }

File: src/bridge/protocol.ts
/**
 * THE WIRE PROTOCOL
 *
 * The browser stand-in speaks this today; the iOS app in Phase 7 speaks the
 * same thing. Keeping it boring and explicit is the point — this is the
 * seam that has to survive a change of client.
 *
 * Text frames carry JSON control messages. Binary frames carry raw PCM
 * (device -> server) or encoded audio (server -> device), tagged by a
 * one-byte header so a binary frame never needs a matching JSON message.
 */

export const BINARY_TAG = {
  /** device -> server: 16 kHz mono s16le microphone audio */
  micPcm: 0x01,
  /** device -> server: JPEG still, in reply to a captureFrame request */
  frameJpeg: 0x02,
  /** server -> device: audio to play (mime declared in the preceding
   *  speakBegin control message) */
  ttsAudio: 0x10,
} as const;

/** device -> server */
export type ClientMessage =
  | {
      t: "hello";
      label: string;
      capabilities: { audioIn: boolean; audioOut: boolean; frames: boolean; display: boolean };
      battery: number | null;
      /** Client-side wake word engine in use, for the session banner. */
      wake: "porcupine" | "push-to-talk" | "none";
    }
  /** Wake word fired, or the user pressed the talk key. Audio follows. */
  | { t: "wake"; at: number; source: "porcupine" | "push-to-talk" }
  /** User released the talk key / client-side VAD saw end of speech. */
  | { t: "endOfSpeech"; at: number }
  /** The user started talking while we were speaking — barge-in (item 03). */
  | { t: "bargeIn"; at: number }
  /** Transcript from client-side recognition, when STT_PROVIDER=browser. */
  | { t: "transcript"; text: string; final: boolean; at: number }
  | { t: "frameError"; reason: string }
  | { t: "battery"; level: number }
  | { t: "playbackDone"; interrupted: boolean }
  | { t: "pong"; at: number };

/** server -> device */
export type ServerMessage =
  | { t: "ready"; persona: { name: string; greeting: string } }
  /** Ask for one still. Reply with a frameJpeg binary frame or frameError. */
  | { t: "captureFrame"; reason: string }
  /** Audio is about to arrive as binary frames. */
  | { t: "speakBegin"; mime: string; utteranceId: string }
  | { t: "speakEnd"; utteranceId: string }
  /** Stop playing immediately and discard buffered audio. */
  | { t: "speakCancel"; utteranceId: string }
  /** For the client to render — what it heard, what it is doing. */
  | { t: "state"; phase: "idle" | "listening" | "thinking" | "speaking"; detail?: string }
  | { t: "heard"; text: string; final: boolean }
  | { t: "said"; text: string }
  | { t: "notice"; level: "info" | "warn" | "error"; text: string }
  | { t: "ping"; at: number };

export function encodeBinary(tag: number, payload: Buffer): Buffer {
  const out = Buffer.allocUnsafe(payload.length + 1);
  out[0] = tag;
  payload.copy(out, 1);
  return out;
}

export function decodeBinary(buf: Buffer): { tag: number; payload: Buffer } | null {
  if (buf.length < 1) return null;
  return { tag: buf[0]!, payload: buf.subarray(1) };
}
